### PR TITLE
CNF-13731: Add HTTP01ChallengeProxy

### PR DESCRIPTION
**Pull request by sebrandon1**
Related to: https://github.com/openshift/enhancements/pull/1773

This is my first attempt at a FeatureGate so I'm expecting this to need a bunch of work.

### Additions to API Specification and Validation:

* Added `HTTP01ChallengeProxy` configuration to the `APIServerSpec` struct, allowing users to enable and configure the HTTP01 challenge proxy for API endpoint certificates. This includes options for `DefaultDeployment` and `CustomDeployment` modes. (`config/v1/types_apiserver.go`, [[1]](diffhunk://#diff-616d67895c3421c2d091662d30ed47b4b6f0b57db9411e29618d22b964ddb9efR71-R76) [[2]](diffhunk://#diff-616d67895c3421c2d091662d30ed47b4b6f0b57db9411e29618d22b964ddb9efR243-R272)

### Feature Gate Registration:

* Registered the new `HTTP01ChallengeProxy` feature gate in `features.go`, enabling it in the `TechPreviewNoUpgrade` scope and providing metadata such as a contact person and enhancement proposal link. (`features/features.go`, [features/features.goR869-R876](diffhunk://#diff-a8b6135d50534471326ea7bcd20e0f5eae25353f7788338060f718128a6a0b34R869-R876))

### Test Cases for Validation:

* Added comprehensive test cases in `HTTP01ChallengeProxy.yaml` to validate the behavior of the `HTTP01ChallengeProxy` configuration. Tests cover valid configurations, invalid modes, and boundary conditions for the `internalPort` field. (`config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml`, [config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yamlR1-R105](diffhunk://#diff-286873dc3e2d6b3790998283ffbed751db492cc94e68c2c7ba3845ee8310fcd6R1-R105))

### CODE REVIEW

File: config/v1/types_apiserver.go [hunk 1]
```diff
@@ -234,6 +240,36 @@ const (
...
+}
+
+// +kubebuilder:validation:MinProperties=1
+type HTTP01ChallengeProxyCustomDeploymentSpec struct {
+	// internalPort specifies the internal port used by the proxy service.
+	// Valid values are 1024-65535. Defaults to 8888.
+	// This port is used to avoid conflicts with other workloads that may require port 8888 on the host.
+	// +kubebuilder:validation:Minimum=1024
+	// +kubebuilder:validation:Maximum=65535
+	// +kubebuilder:default=8888
```
Comment (everettraven) [inline config/v1/types_apiserver.go]: If they are specifying a custom deployment, should we allow this to default? Are we always going to want to default to `8888`? Setting it this way means we do not reserve the right to change it in the future, is there any chance we would need to change this default in the future?

Should this instead be required with no default?

Looks like @JoelSpeed had the same comment in the EP here: https://github.com/openshift/enhancements/pull/1773#discussion_r2245851448

Comment (sebrandon1) [reply config/v1/types_apiserver.go]: Adjusted in the PR.  Thanks!

**Review by everettraven — COMMENTED** [review]
I'm surprised to see so many protobuf changes here. Something seems off with this many changes being made.

@JoelSpeed is probably more familiar with what might be going on here than I.

Because you are making changes to a CRD, you should be able to run `PROTO_OPTIONAL=true make update` to make generation updates without touching the protobuf stuff

**Comment by JoelSpeed** [comment]
Proto changes appear to be unrelated, verify thinks changing them is wrong and is suggesting to change them back

Possibly running with a newer or incompatible version?

**Comment by sebrandon1** [comment]
Okay I started over and re-ran it with `PROTO_OPTIONAL=true make update` like you suggested and it seems much more manageable now.  I'll take whatever suggestions I can get 😄

File: features/features.go [hunk 2]
```diff
@@ -872,4 +872,12 @@ var (
 						enhancementPR("https://github.com/openshift/enhancements/pull/1785").
 						enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
 						mustRegister()
+
+	FeatureGateHTTP01ChallengeProxy = newFeatureGate("HTTP01ChallengeProxy").
+					reportProblemsToJiraComponent("kube-apiserver").
+					contactPerson("sebrandon1").
+					productScope(ocpSpecific).
+					enhancementPR("https://github.com/openshift/enhancements/pull/1773").
+					enableIn(configv1.TechPreviewNoUpgrade).
```
Comment (everettraven) [inline features/features.go]: Add `configv1.DevPreviewNoUpgrade` here as well.

**Comment by everettraven** [comment]
> Okay I started over and re-ran it with `PROTO_OPTIONAL=true make update` like you suggested and it seems much more manageable now. I'll take whatever suggestions I can get 😄

Looks like the actual API changes are missing now

File: features/features.go (hunk 2 above)
Comment (sebrandon1) [reply features/features.go]: Changed in features.go.

File: config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    - name: Should reject CustomDeployment mode with port above maximum 65536
+      initial: |
+        apiVersion: config.openshift.io/v1
+        kind: APIServer
+        spec:
+          http01ChallengeProxy:
+            mode: CustomDeployment
+            customDeployment:
+              internalPort: 65536
+      expectedError: "spec.http01ChallengeProxy.customDeployment.internalPort: Invalid value: 65536: must be less than or equal to 65535"
```
Comment (JoelSpeed) [inline config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml]: Nit, final character should be a new line

File: config/v1/types_apiserver.go [hunk 4]
```diff
@@ -68,6 +68,12 @@ type APIServerSpec struct {
 	// +optional
 	// +kubebuilder:default={profile: Default}
 	Audit Audit `json:"audit"`
+	// http01ChallengeProxy contains configuration for the HTTP01 challenge proxy
+	// that redirects traffic from the API endpoint on port 80 to ingress routers.
+	// This enables cert-manager to perform HTTP01 ACME challenges for API endpoint certificates.
+	// +openshift:enable:FeatureGate=HTTP01ChallengeProxy
+	// +optional
+	HTTP01ChallengeProxy *HTTP01ChallengeProxySpec `json:"http01ChallengeProxy,omitempty"`
```
Comment (JoelSpeed) [inline config/v1/types_apiserver.go]: Zero value for this should not be valid, so this can be `omitzero`
```suggestion
	HTTP01ChallengeProxy HTTP01ChallengeProxySpec `json:"http01ChallengeProxy,omitempty,omitzero"`
```

Linter isn't currently reporting this because of the defaulting of the `mode` field

File: config/v1/types_apiserver.go [hunk 5]
```diff
@@ -234,6 +240,36 @@ const (
...
 
+// +union
+// +kubebuilder:validation:XValidation:rule="self.mode == 'CustomDeployment' ? has(self.customDeployment) : !has(self.customDeployment)",message="customDeployment is required when mode is CustomDeployment and forbidden otherwise"
+type HTTP01ChallengeProxySpec struct {
+	// mode controls whether the HTTP01 challenge proxy is active and how it should be deployed.
+	// DefaultDeployment enables the proxy with default configuration.
+	// CustomDeployment enables the proxy with user-specified configuration.
+	// +kubebuilder:validation:Enum=DefaultDeployment;CustomDeployment
+	// +kubebuilder:default=DefaultDeployment
+	// +optional
```
Comment (JoelSpeed) [inline config/v1/types_apiserver.go]: Why are we defaulting this? I don't think it really makes sense

You've made the parent of this field optional and omitted by default, so the only way to trigger this default is if someone were to try and set `http01ChallengProxy: {}`, which, is an odd behaviour?

I think make this required and remove the default

File: config/v1/types_apiserver.go [hunk 6]
```diff
@@ -234,6 +240,36 @@ const (
 	EncryptionTypeKMS EncryptionType = "KMS"
 )
 
+// +union
+// +kubebuilder:validation:XValidation:rule="self.mode == 'CustomDeployment' ? has(self.customDeployment) : !has(self.customDeployment)",message="customDeployment is required when mode is CustomDeployment and forbidden otherwise"
+type HTTP01ChallengeProxySpec struct {
+	// mode controls whether the HTTP01 challenge proxy is active and how it should be deployed.
+	// DefaultDeployment enables the proxy with default configuration.
+	// CustomDeployment enables the proxy with user-specified configuration.
+	// +kubebuilder:validation:Enum=DefaultDeployment;CustomDeployment
```
Comment (JoelSpeed) [inline config/v1/types_apiserver.go]: For enums, we generally create a type alias and constants to represent the different values, can we do that here too please

File: config/v1/types_apiserver.go [hunk 7]
```diff
@@ -234,6 +240,36 @@ const (
...
+
+// +kubebuilder:validation:MinProperties=1
+type HTTP01ChallengeProxyCustomDeploymentSpec struct {
+	// internalPort specifies the internal port used by the proxy service.
+	// Valid values are 1024-65535.
+	// This port must be specified to avoid conflicts with other workloads on the host.
+	// +kubebuilder:validation:Minimum=1024
+	// +kubebuilder:validation:Maximum=65535
+	// +required
+	InternalPort int32 `json:"internalPort"`
```
Comment (JoelSpeed) [inline config/v1/types_apiserver.go]: Add `omitempty` here

File: config/v1/types_apiserver.go [hunk 8]
```diff
@@ -234,6 +240,36 @@ const (
...
+	Mode string `json:"mode,omitempty"`
+
+	// customDeployment contains configuration options when mode is CustomDeployment.
+	// This field is only valid when mode is CustomDeployment.
+	// +optional
+	// +unionMember
+	CustomDeployment HTTP01ChallengeProxyCustomDeploymentSpec `json:"customDeployment,omitzero,omitempty"`
+}
+
+// +kubebuilder:validation:MinProperties=1
```
Comment (JoelSpeed) [inline config/v1/types_apiserver.go]: Not strictly needed since you have a required field, the zero value `{}` would fail validation on the required field not being present, and `interalPort: 0` would also fail validation because `0` is not valid.

You can leave this though if you prefer to keep it

File: config/v1/types_apiserver.go [hunk 9]
```diff
@@ -234,6 +240,45 @@ const (
...
+}
+
+type HTTP01ChallengeProxyCustomDeploymentSpec struct {
+	// internalPort specifies the internal port used by the proxy service.
+	// Valid values are 1024-65535.
+	// When not specified for CustomDeployment mode, users should ensure their chosen port
+	// does not conflict with other workloads on the host.
+	// +kubebuilder:validation:Minimum=1024
+	// +kubebuilder:validation:Maximum=65535
+	// +optional
```
Comment (JoelSpeed) [inline config/v1/types_apiserver.go]: I see you changed this to optional, it shouldn't be optional, the error you saw was a false positive, lets revert this back to required please

Comment (sebrandon1) [reply config/v1/types_apiserver.go]: Whoops yes you are correct.  I'll adjust.

File: config/v1/types_apiserver.go [hunk 10]
```diff
@@ -234,6 +240,45 @@ const (
...
+	// CustomDeployment enables the proxy with user-specified configuration.
+	// +required
+	// +unionDiscriminator
+	Mode HTTP01ChallengeProxyMode `json:"mode,omitempty"`
+
+	// customDeployment contains configuration options when mode is CustomDeployment.
+	// This field is only valid when mode is CustomDeployment.
+	// +optional
+	// +unionMember
+	CustomDeployment *HTTP01ChallengeProxyCustomDeploymentSpec `json:"customDeployment,omitzero,omitempty"`
```
Comment (JoelSpeed) [inline config/v1/types_apiserver.go]: I don't think this needs to be a pointer now (I suspect linter failure agrees)

Comment (sebrandon1) [reply config/v1/types_apiserver.go]: Pointer removed!  Let's see how the linter likes it.

**Comment by JoelSpeed** [comment]
API types look good, but it looks like some of the test cases are expecting things that aren't true (optional port), so lets get those fixed up and please squash the various fixups

File: config/v1/types_apiserver.go [hunk 11]
```diff
@@ -208,7 +214,7 @@ type APIServerEncryption struct {
 	// +openshift:enable:FeatureGate=KMSEncryptionProvider
 	// +unionMember
 	// +optional
-	KMS *KMSConfig `json:"kms,omitempty"`
+	KMS KMSConfig `json:"kms,omitzero,omitempty"`
 }
```
Comment (everettraven) [inline config/v1/types_apiserver.go:218]: Just wanting to verify - these changes are specifically because of linter issues that you were fixing?

File: config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
```
Comment (everettraven) [inline config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml]: integration check in CI is failing because this expected error doesn't match what it actually recieved:
```
spec.http01ChallengeProxy.customDeployment.internalPort: Invalid value: 65536: spec.http01ChallengeProxy.customDeployment.internalPort in body should be less than or equal to 65535
```

File: config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+          http01ChallengeProxy:
+            mode: DefaultDeployment
+      expected: |
+        apiVersion: config.openshift.io/v1
+        kind: APIServer
+        spec:
+          audit:
+            profile: Default
+          http01ChallengeProxy:
+            mode: DefaultDeployment
```
Comment (everettraven) [inline config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml:22]: Not sure why, but in the integration check in CI, it looks like this is getting back:
```yaml
apiVersion: config.openshift.io/v1 
kind: APIServer
spec:
  audit:
    profile: Default
```
(missing the `http01ChallengeProxy` configuration)

Comment (everettraven) [reply config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml:22]: Ah, I found it - it is thinking these tests apply for the default CRD as well which this field isn't present in.

You are missing something like this at the top of this test file to signal that this should only apply for featuresets where this feature-gate is enabled:
https://github.com/openshift/api/blob/4a165b214722ff53fbdac1d250f62d4b0bb2bd3d/config/v1/tests/authentications.config.openshift.io/ExternalOIDC.yaml#L4-L5

Comment (sebrandon1) [reply config/v1/tests/apiservers.config.openshift.io/HTTP01ChallengeProxy.yaml:22]: Added:

```
featureGates:
  - HTTP01ChallengeProxy
  ```

**Comment by everettraven** [comment]
False positive on new required field with optional parent.

/override ci/prow/verify-crd-schema

File: config/v1/types_apiserver.go [hunk 14]
```diff
@@ -114,6 +120,7 @@ type Audit struct {
 	// If unset, the 'Default' profile is used as the default.
 	//
 	// +kubebuilder:default=Default
+	// +optional
```
Comment (everettraven) [inline config/v1/types_apiserver.go:122]: Unrelated change? Still one the linter is trying to get you to make?

File: config/v1/types_apiserver.go [hunk 15]
```diff
@@ -146,7 +153,7 @@ type AuditCustomRule struct {
 	// If unset, the 'Default' profile is used as the default.
 	//
 	// +required
-	Profile AuditProfileType `json:"profile"`
+	Profile AuditProfileType `json:"profile,omitempty"`
```
Comment (everettraven) [inline config/v1/types_apiserver.go]: Unrelated change? Still one the linter is trying to get you to make?

File: config/v1/types_apiserver.go (hunk 14 above)
Comment (sebrandon1) [reply config/v1/types_apiserver.go:122]: Reverted as it is unrelated.

File: config/v1/types_apiserver.go (hunk 15 above)
Comment (sebrandon1) [reply config/v1/types_apiserver.go]: Reverted as it is unrelated.

**Comment by JoelSpeed** [comment]
Changes here LGTM, there are however some code generation issues by the looks of it

**Comment by everettraven** [comment]
/approve
/lgtm

verify-crd-schema failing on a false positive. New required fields are all parented by optional fields.

/override ci/prow/verify-crd-schema

Holding this PR on merge of enhancement proposal. @JoelSpeed Feel free to overrule me here if holding on EP merge is not necessary.
/hold

**Comment by JoelSpeed** [comment]
Regarding whether this should merge or not, I currently haven't seen any of the listed EP approvers chime on the EP for review, nor anyone from the kube-apiserver-operator owners file. Since the EP relies on that team accepting a change into their operator, I'd expect to see at least some signal that they're happy with the direction before we merge the API
